appendix-keyboard.tex

```diff
@@ -148,14 +148,33 @@ \section{Synthetic Key Events}
 
 \section{Keyboard LED Control}
 
-The LEDs on the MEGA65's keyboard are normally controlled automatically
-by the system.  However, it is also possible to place them under
-user control.  This is activated by setting bit 7 of \$D61D (decimal 54813).
-The lower bits indicate which keyboard register to set.  Registers 0 through 11
-correspond to the red, green and blue channels of the four LEDs.  The intensity
-of an LED is set by setting \$D61E (decimal 54814) to the desired intensity.
+The LEDs on the MEGA65's keyboard are normally controlled automatically by the
+system.  However, it is also possible to place them under user control.  This
+is activated by setting bit 7 (decimal 128) of \$D61D (decimal 54813).  The
+lower bits indicate which keyboard LED to set.  Values 0 through 11 correspond
+to the red, green and blue channels of the four LEDs. The table below shows the
+specific values:
 
-To return the keyboard LEDs to hardware control, clear bit 7 of \$D61D.
+\begin{adjustwidth}{}{-2cm}
+\begin{description}[align=left,labelwidth=0.2cm]
+\item[ 0] left-half of DRIVE LED, RED
+\item[ 1] left-half of DRIVE LED, GREEN
+\item[ 2] left-half of DRIVE LED, BLUE
+\item[ 3] right-half of DRIVE LED, RED
+\item[ 4] right-half of DRIVE LED, GREEN
+\item[ 5] right-half of DRIVE LED, BLUE
+\item[ 6] left-half of POWER LED, RED
+\item[ 7] left-half of POWER LED, GREEN
+\item[ 8] left-half of POWER LED, BLUE
+\item[ 9] right-half of POWER LED, RED
+\item[10] right-half of POWER LED, GREEN
+\item[11] right-half of POWER LED, BLUE
+\end{description}
+\end{adjustwidth}
+
+In other words, the intensity of an LED can be set by first setting the desired

```

Comment:
Alternative wording:
```
Register \$D61E (decimal 54814) needs to be used to specify the intensity value
(0..255).  To change the brightness of a specific LED, first set the desired
intensity value into \$D61E and then set \$D61D to 128 + x, where x is a value
from the table above.
```

Reply:
So it's slightly more tricky than this:
Whatever value is in $D61E gets written to whatever register is selected in $D61D. So if you change $D61E before $D61D, you will cause whichever register is currently selected by $D61D. So to do it safely, you need to set $D61D first, then $D61E. That will cause a brief glitch. To do it 100% safely, set $D61D to an invalid register, e.g., $FF, then set $D61E to the new intensity, then finally set $D61D to the register you wish to set.  Then there has to be a bit of a time delay if you want to set multiple ones, as only one register can be set per keyboard cycle.  A ~1ms interval is almost certainly safe. Some experimentation will be required to confirm the exact limit. There is a register that sets the keyboard scan rate that affects it. It's up to you whether we want to document that.

Reply:
> It's up to you whether we want to document that.

Yes we do. We also want assembly programmers to be able to use this.
Please have a look if 5e128fb980f3da10739ecc1953554552ac42686e is accurate.

Reply:
> So if you change $D61E before $D61D, you will cause whichever register is currently selected by $D61D. So to do it safely, you need to set $D61D first, then $D61E. That will cause a brief glitch.

BTW, thanks for letting me know what's happening under the hood here.

I would have designed that a little different by using the act of writing to $D61D, $D61E to copy the values into hidden registers. Then when the keyboard logic comes by and it sees both values set (and only then) it acts on them and then resets them both. That way no glitches can ever happen and the consumption of new LED configurations is effectively decoupled from the keyboard scan rate.

Reply:
Good idea to make it latch.  I'll add that in.  It also turns out that $D61E was write-only, so I'll fix that at the same time.  Tracked under https://github.com/MEGA65/mega65-core/issues/238